### ISSUE TITLE
sort_values(), sort_index(), take a `key` argument?

### ISSUE THREAD

**Issue by flexatone** [InvestmentSystems/static-frame#29]

**Comment by flexatone** [comment]
key would have to be a function called on the entire container and returning a new container with the appropriate shape, to be then sorted and used to derive the sort order of the original container.

**Comment by ForeverWintr** [comment]
Here is an example case where such a feature would be useful. Say I have the following frame:

```python
f = sf.Frame.from_dict({('a', 'x'): tuple('AABB'), ('a', 'y'): tuple('AACC'), ('c', 'x'): tuple('CCBB'), ('c', 'y'): tuple('CCCC'), ('d', 'x'): tuple('DDBB'), ('d', 'y'): tuple('DDCC')} )   
f = f.relabel(columns=sf.IndexHierarchy.from_labels(f.columns))

<Frame>
<IndexHierarchy> a     a     c     c     d     d     <<U1>
                 x     y     x     y     x     y     <<U1>
<Index>
0                A     A     C     C     D     D
1                A     A     C     C     D     D
2                B     C     B     C     B     C
3                B     C     B     C     B     C
<int64>          <<U1> <<U1> <<U1> <<U1> <<U1> <<U1>
```

And I want to sort its columns so that the 'c' columns appear before 'a' and 'd'. 

I can accomplish this using python's `sorted` with a `key` argument like so:

```python
def key(element):
    '''Custom sort key, spelled our for clarity.'''
    x, y = element
    if x == 'c':
        return ('_', y)
    return element

new_order = sorted(f.columns, key=key)
new_f = f.reindex(columns=sf.IndexHierarchy.from_labels(new_order))
```

However, it would be more convenient to simply pass my key function to `Frame.sort_columns` directly.

Note that contrary to your comment above, python's sort key does not operate on the entire container. It is called for each element and returns a new element that is used only for sorting.

**Comment by flexatone** [comment]
Pandas has added a key argument to `sort_values` and `sort_index` as of 1.1.0; there, the key is applied not per element, but per 1D vector (either Series or Index levels, it seems).

https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.sort_values.html
https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.sort_index.html

**Comment by gdevanla** [comment]
This would be nice feature to have. To motivate this, here is another use case:

```
In [96]: f=parse('f(Fg)|v(str, str)|s(3,3)')

In [97]: f
Out[97]: 
<FrameGO>
<IndexGO> 0     1     2     <int64>
<Index>
0         zjZQ  zaji  ztsv
1         zO5l  zJnC  zUvW
2         zEdH  zDdR  zkuW
<int64>   <<U4> <<U4> <<U4>

In [98]: order = {'zUvW': 0, 'zkuW': 1, 'ztsv': 2} # I would like to sort `2` based on some pre-determined order listed here

In [99]: f.sort_values(2,  key_=lambda z: order[z])   # Note: This does not work right now but if it worked, the example expects the `key_` argument to work as `key` argument of standard Python.

```

If we were to follow Pandas and perform this operation, then I think one of the approaches to achieve what I need would be as follows:

```
order = {'zUvW': 0, 'zkuW': 1, 'ztsv': 2}
def sort_order(s: sf.Series):
    s.map(order)
f.sort_values(2, key_=sort_order)

```
I think having `key` support vectoriz-ed operations is sufficient to support special cases of sorting by applying functions to individual values.

**Comment by flexatone** [comment]
Thanks, @gdevanla , for this example. Given these examples, I think that having the key be a `Series` or `Frame`-level processor offers the best opportunity for performance, as well as still providing a path-way to element-level processing if needed. I will explore the implementation for 0.7.15.

**Comment by brandtbucher** [comment]
Just came across another need for this: sorting columns, where the column labels are bond maturities (`'3Y'`, `'30Y'`, `'3M'`, etc...). Needs a custom key to get it right.

**Comment by flexatone** [comment]
This features has been completed for all sorting routines and will be available as part of the 0.8.0 release.

In general, the `key` function must take as input the type of the container (post selection, in the case of the `Frame`) and return either the same type, a lower dimensional type (i.e., a `Series`), or an np.ndarray. For example, if we select column to sort, the `key` function must be a `Series` processor:

```python
>>> f = ff.parse('f(Fg)|v(str, str)|s(3,3)')                                              
>>> f                                                                                     
<FrameGO>
<IndexGO> 0     1     2     <int64>
<Index>
0         zjZQ  zaji  ztsv
1         zO5l  zJnC  zUvW
2         zEdH  zDdR  zkuW
<int64>   <<U4> <<U4> <<U4>
>>> order = {'zUvW': 0, 'zkuW': 1, 'ztsv': 2}                                             
>>> f.sort_values(2, key=lambda s:s.iter_element().map_all(order))                        
<FrameGO>
<IndexGO> 0     1     2     <int64>
<Index>
1         zO5l  zJnC  zUvW
2         zEdH  zDdR  zkuW
0         zjZQ  zaji  ztsv
<int64>   <<U4> <<U4> <<U4>
```
When using `sort_columns`, the ` key` must process the index object. Note that in 0.8.0 `iter_label().apply()` now returns an np.ndarray, facilitating this application.

```python
>>> f = sf.Frame.from_dict({('a', 'x'): tuple('AABB'), ('a', 'y'): tuple('AACC'), ('c', 'x'): tuple( 
... 'CCBB'), ('c', 'y'): tuple('CCCC'), ('d', 'x'): tuple('DDBB'), ('d', 'y'): tuple('DDCC')} )                             
>>> f = f.relabel(columns=sf.IndexHierarchy.from_labels(f.columns))                                                                                                                                                                       
>>> f.sort_columns(key=lambda i:i.iter_label().apply(lambda x:0 if x[0] == 'c' else 1))                                         
<Frame>
<IndexHierarchy> c     c     a     a     d     d     <<U1>
                 x     y     x     y     x     y     <<U1>
<Index>
0                C     C     A     A     D     D
1                C     C     A     A     D     D
2                B     C     B     C     B     C
3                B     C     B     C     B     C
<int64>          <<U1> <<U1> <<U1> <<U1> <<U1> <<U1>
```
